src/components/ITSBStore/itsbGraph.js

```diff
@@ -0,0 +1,149 @@
+import createGraph from 'ngraph.graph';
+import { 
+  normalizeNode,  
+  filterByTime,
+  groupBy, 
+  sortWaypoints,
+  splitItinerary
+} from './utils';
+
+/**
+ * A domain graph model for ITSB with the following 
+ * types of nodes:
+ * 
+ * - author (id, name)
+ * - place (GeoJSON/LP feature)
+ * - waypoint (one stop on the itinerary, linked to 
+ *   author, place, citations, and scoped by time
+ * 
+ * Edges:
+ * - waypoints are linked to author
+ * - waypoints are linked to places
+ * - waypoints are (directionally!) linked to waypoints
+ */
+export class ITSBGraph {
+
+  constructor() {
+    this.graph = createGraph();
+  }
+
+  init = (authors, places, itineraries) => {
+    this.graph.clear();
+
+    this.graph.beginUpdate();
+
+    // Authors and places can be treated the same way
+    const authorNodes = authors.map(normalizeNode);
+    const placeNodes = places.map(normalizeNode);
+
+    [...authorNodes, ...placeNodes ].forEach(n => 
+      this.graph.addNode(n.id, n));
+
+    // Itineraries are split to waypoint nodes
+    const waypointNodes = itineraries.reduce((waypoints, it) => {
+      return [...waypoints, ...splitItinerary(it)]
+    }, []).filter(wp => { 
+      const isValid = this.exists(wp.author, wp.place);
+
+      if (!isValid)
+        console.warn('Skipping unconnected node', wp);
+
+      return isValid;
+    });
+
+    waypointNodes.forEach(wp => this.graph.addNode(wp.id, wp));
+
+    // Link waypoints to authors
+    waypointNodes.forEach(wp => {
+      const sourceId = wp.id;
+      const targetId = wp.author;
+      this.graph.addLink(sourceId, targetId, { relation: 'visitedBy' });
+    });
+
+    // Link waypoints to places
+    waypointNodes.forEach(wp => {
+      const sourceId = wp.id;
+      const targetId = wp.place;
+      this.graph.addLink(sourceId, targetId, { relation: 'locatedAt' });
+    });
+
+    // Link waypoints in sequence
+    waypointNodes.reduce((previous, wp) => {
+      if (previous && previous.author === wp.author) {
+        const sourceId = wp.id;
+        const targetId = previous.id;
+        this.graph.addLink(sourceId, targetId, { relation: 'previous'});
+      }
+
+      return wp;
+    }, null);
```

Comment:
And this sequence is just whatever sequence results from the JSON --> `[...waypoints, ...splitItinerary(it)]` --> filtering out invalid nodes, right? I wonder if the itinerary splitting could be what causes some of them to get out of temporal order as you mentioned on Slack?

Reply:
Yes, exactly. This will blindly rely on the sequence in the original JSON. The "splitting" will actually just do a map on the "body" array, i.e. order should be preserved, there, too.